### PR TITLE
feat(repository): apply retention policies server-side

### LINKED ISSUE #2023: Snapshot migrate or create errors when using server with APPEND only ACL for snapshots
Hello,

When using kopia server with this policies:
```
id:d62f711dd6c06759bb461c6b236b1fb0 user:*@* access:APPEND target:type=content
id:a0517d9124156b1cc86f83d09a0492e0 user:*@* access:READ target:type=policy,policyType=global
id:f1458d3121e5deefbb8cd6157b9aa820 user:*@* access:READ target:type=policy,hostname=OWN_HOST,policyType=host
id:f97b833fef122f824fd8be61c02fe0de user:*@* access:READ target:type=user,username=OWN_USER@OWN_HOST
id:dc3379322d56c3d6a29aaa1e14637240 user:*@* access:READ target:type=policy,username=OWN_USER,hostname=OWN_HOST
id:57691fe221d611e5df75feb9090f8c6e user:*@* access:APPEND target:type=snapshot,hostname=OWN_HOST,username=OWN_USER
```

On `snapshot migrate` or `snapshot create` error occurs because kopia tries to apply retention policy, but ACL denies it.
In my case it occured during checkpoint process, and investigation showed that reason is possible [here](https://github.com/kopia/kopia/blob/master/snapshot/snapshotfs/upload.go#L433).

It will be nice to have a flag which will totally disable retention policy apply for commands mentioned above.

### PR DESCRIPTION
**Pull request by jkowalski**
This allows append-only snapshots where the client can never delete arbitrary manifests and policies are maintained on the server.

The client only needs permissions to create snapshots in a given source (APPEND), which automatically gives them permission to invoke the server-side method for their own snapshots only. The client still needs `FULL` access to delete snapshots.

Added `--overwrite` method to `kopia server acl add` that allows reducing access level for a combination of user/target.

Fixes #2023

### CODE REVIEW

**Comment by jkowalski** [comment]
note this needs more work before it can get merged, we need to add a bunch of tests, in particular make sure this support #3227.

File: cli/command_acl_add.go [hunk 1]
```diff
@@ -11,16 +11,18 @@ import (
...
+	level     string
+	overwrite bool
 }
 
 func (c *commandACLAdd) setup(svc appServices, parent commandParent) {
 	cmd := parent.Command("add", "Add ACL entry")
 	cmd.Flag("user", "User the ACL targets").Required().StringVar(&c.user)
 	cmd.Flag("target", "Manifests targeted by the rule (type:T,key1:value1,...,keyN:valueN)").Required().StringVar(&c.target)
 	cmd.Flag("access", "Access the user gets to subject").Required().EnumVar(&c.level, acl.SupportedAccessLevels()...)
+	cmd.Flag("overwrite", "Overwrite existing rule with the same user and targe").BoolVar(&c.overwrite)
```
Comment (Gui13) [inline cli/command_acl_add.go:25]:
```suggestion
	cmd.Flag("overwrite", "Overwrite existing rule with the same user and target").BoolVar(&c.overwrite)
```

File: internal/server/api_manifest.go [hunk 2]
```diff
@@ -123,3 +126,47 @@ func handleManifestCreate(ctx context.Context, rc requestContext) (interface{},
...
+	}
+
+	usernameAtHostname, _, _ := rc.req.BasicAuth()
+
+	parts := strings.Split(usernameAtHostname, "@")
+	if len(parts) != 2 { //nolint:gomnd
+		return nil, requestError(serverapi.ErrorMalformedRequest, "malformed username")
+	}
+
+	// only allow users with apply retention policy if they have permission to add snapshots
```
Comment (Gui13) [inline internal/server/api_manifest.go:149]:
```suggestion
	// only allow users to apply retention policy if they have permission to add snapshots
```

File: internal/server/api_manifest.go [hunk 3]
```diff
@@ -123,3 +126,47 @@ func handleManifestCreate(ctx context.Context, rc requestContext) (interface{},
...
+	// for a particular parth.
```
Comment (Gui13) [inline internal/server/api_manifest.go:150]:
```suggestion
	// for a particular path.
```

File: internal/server/grpc_session.go [hunk 4]
```diff
@@ -440,6 +445,47 @@ func handlePrefetchContentsRequest(ctx context.Context, rep repo.Repository, aut
...
+
+	parts := strings.Split(usernameAtHostname, "@")
+	if len(parts) != 2 { //nolint:gomnd
+		return errorResponse(errors.Errorf("invalid username@hostname: %q", usernameAtHostname))
+	}
+
+	username := parts[0]
+	hostname := parts[1]
+
+	// only allow users with apply retention policy if they have permission to add snapshots
```
Comment (Gui13) [inline internal/server/grpc_session.go:460]:
```suggestion
	// only allow users to apply retention policy if they have permission to add snapshots
```

File: snapshot/policy/expire.go [hunk 5]
```diff
@@ -7,11 +7,24 @@ import (
...
+func ApplyRetentionPolicy(ctx context.Context, rep repo.RepositoryWriter, sourceInfo snapshot.SourceInfo, reallyDelete bool) ([]manifest.ID, error) {
+	// it is desired to not allow snapshots to be deleted by repository clients,
+	// while still maintain the ability to apply snapshot retention policies server-side.
+	if remote, ok := rep.(repo.RemoteRetentionPolicy); ok {
+		if sourceInfo.UserName == rep.ClientOptions().Username && sourceInfo.Host == rep.ClientOptions().Hostname {
+			// for repository clients, apply retention policy on the server.
+			log(ctx).Debug("applying retention policy on the server")
+
+			//nolint:wrapcheck
+			return remote.ApplyRetentionPolicy(ctx, sourceInfo.Path, reallyDelete)
```
Comment (Gui13) [inline snapshot/policy/expire.go:24]: That's the crux, thanks for that @jkowalski ! :heart:

**Pull request merged**